### PR TITLE
Improve build.rs and some more. 

### PR DESCRIPTION
**Pull request by Sh3Rm4n**
I had some ideas about the newly added build.rs and touched some documentation as well. 

What do you think about this change, @Piroro-hs?

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -10,40 +19,69 @@ fn main() {
...
-Please select one of the following (`x` denotes any character in [a-z]):
+    // get all device variants from the metadata
+    let metadata = MetadataCommand::new().exec().unwrap();
+    let mut device_variants = metadata.root_package().unwrap().features.clone();
+    device_variants.retain(|_, dependend_features| {
+        dependend_features
+            .iter()
+            .any(|feature| *feature == "device-selected")
+    });
+    let mut device_variants = device_variants.keys().cloned().collect::<Vec<String>>();
```
Comment (Piroro-hs) [inline build.rs:44]: One `filter_map` call is enough?

**Review by Piroro-hs — CHANGES_REQUESTED** [review]
Ensuring user select exactly one device is great I think :)

File: build.rs (hunk 1 above)
Comment (Piroro-hs) [reply build.rs:44]:
```rust
    let features = metadata.root_package().unwrap().features.iter();
    let mut device_variants: HashSet<String> = features
        .filter_map(|(feature, dependend_features)| {
            dependend_features
                .iter()
                .any(|dependend_feature| dependend_feature == "device-selected")
                .then(|| feature.clone())
            // `bool::then` is only available for Rust >= 1.50
            // match dependend_features
            //     .iter()
            //     .any(|feature| feature == "device-selected")
            // {
            //     true => Some(feature.clone()),
            //     false => None,
            // }
        })
        .collect();
```

**Comment by Piroro-hs** [comment]
How about changing those `eprintln!`s like this?
```rust
    eprintln!(
        "I've just found Rust's string literal support line breaking.\n\
        \n\
        We can keep indentation!!!."
    );
```

**Comment by Sh3Rm4n** [comment]
Oh, I've searched for a variation to keep the indentation as well. :smile: Thanks for the tip  :+1:

File: build.rs [hunk 2]
```diff
@@ -10,40 +19,69 @@ fn main() {
...
+        .collect();
 
-Example: The STM32F3Discovery board has a STM32F303VCT6 chip.
+    // check if exactly one device was selected
+    if device_variants
+        .iter()
+        .filter(|device| selected_features.contains(&device.to_ascii_uppercase()))
+        .count()
+        != 1
+    {
```
[the remark this answers is not in the document]
Comment (Sh3Rm4n) [reply build.rs:57]: I really like this idea. I should look more into the `Set` types.

[the remark this answers is not in the document]
Comment (Sh3Rm4n) [reply build.rs:57]: Hm. I think I'll go for a [`Union`](https://doc.rust-lang.org/std/collections/hash_set/struct.Union.html) instead. It's essentially the same, but more obvious. I had to look up, what the result of a `BitOr` was between to `HashSet` even though it's pretty clear in hindsight.

[the remark this answers is not in the document]
Comment (Sh3Rm4n) [reply build.rs:57]: [`intersection`](https://doc.rust-lang.org/std/collections/hash_set/struct.HashSet.html#method.intersection) is what I was wanting to go for.

File: build.rs (hunk 1 above)
Comment (Sh3Rm4n) [reply build.rs]: Bumping MSVR is not a big deal, so I'll go for the `bool::then` method.

File: build.rs [hunk 3]
```diff
@@ -10,40 +20,75 @@ fn main() {
...
+    if !cfg!(feature = "device-selected") && cfg!(feature = "direct-call-deprecated") {
+        eprintln!(
+            "The (device-)feature you selected is deprecated, because it was split up into sub-devices.\n\
+            \n\
+            Example: The STM32F3Discovery board has a STM32F303VCT6 chip.\n\
+            You probably used to use `stm32f303` but now functionalities for the sub-device were added.\n\
+            In this case replace it with `stm32f303xc` to make your code build again.\n\
+            \n\
+            For more information, see \
+            \x1b]8;;https://github.com/stm32-rs/stm32f3xx-hal#selecting-the-right-chip\x1b\\README\
```
Comment (Piroro-hs) [inline build.rs:33]:
```suggestion
            \x1b]8;;https://github.com/stm32-rs/stm32f3xx-hal#selecting-the-right-chip\x1b\\README \
```

File: build.rs [hunk 4]
```diff
@@ -10,40 +20,75 @@ fn main() {
...
-You probably used to use `stm32f303` but now functionalities for the sub-device were added.
-In this case replace it with `stm32f303xc` to make your code build again.
+    // get all device variants from the metadata
+    let metadata = MetadataCommand::new().exec().unwrap();
+    let device_variants: HashSet<String> = metadata
+        .root_package()
+        .unwrap()
+        .features
+        .iter()
+        .filter_map(|(feature, dependend_features)| {
```
Comment (Piroro-hs) [inline build.rs:46]:
```suggestion
        .filter_map(|(feature, dependent_features)| {
```

**Review by Piroro-hs — COMMENTED** [review]
Found some typo

**Comment by Sh3Rm4n** [comment]
Thanks for your review @Piroro-hs

**Pull request merged**